SRF inspection scripts and helper functions

lispandfound opened the pull request:
I wrote these to plot SRF data for the type-5 verification, but these could be very useful for anyone inspecting an SRF. Requires #4 to work correctly.

AndrewRidden-Harper commented on source_modelling/scripts/plot_mw_contributions.py:
```diff
@@ -0,0 +1,98 @@
+
+
+def plot_mw_contributions(
+    srf_ffp: Annotated[
```
noticed that this function doesn't have a docstring

AndrewRidden-Harper commented:
these will be useful tools

lispandfound commented:
@joelridden wanted me to add script outputs, I've gone one better and added a wiki page with a reasonably detailed answer to the question "How do I plot x from an SRF?" We can hopefully point PhD students to this page in the future.

joelridden commented on source_modelling/scripts/plot_mw_contributions.py:
```diff
@@ -0,0 +1,98 @@
+    plt.savefig(output_ffp, dpi=dpi)
+
+
+def main():
+    typer.run(plot_mw_contributions)
+
+
+if __name__ == "__main__":
+    main()
```
If you have just the 1 function in the script you can use the app and decorator and just call app instead of a new function main(), does allow for expansion for multiple functions if needed but unsure if it's "better" or not. When you do this you dont need to specify the function anyways as it's smart and just calls the only function in the file.

```python
app = typer.Typer()

@app.command(
    help="Some Function"
)
def some_function():
    pass

if __name__ == "__main__":
    app()```

lispandfound replied:
Yeah the distinction is actually for the `pyproject.toml` file so that I can create the `plot-*` scripts. It requires an entrypoint function that takes no arguments to call when the user types e.g. `plot-srf ...`. So I added the main function. The entrypoint only has to be `Callable`, however, so I could use the `app` variable instead. Maybe I will do this.

joelridden commented on source_modelling/scripts/plot_rakes.py:
```diff
@@ -0,0 +1,75 @@
+    dpi: Annotated[
+        float, typer.Option(help="Plot output DPI (higher is better).")
+    ] = 300,
+    title: Annotated[str, typer.Option(help="Plot title to use.")] = "Title",
```
Maybe the default title should be None? And then if so just doesnt have a title?

lispandfound replied:
Yeah! Good idea

joelridden commented on source_modelling/scripts/plot_srf.py:
```diff
@@ -0,0 +1,233 @@
+        fill="white",
+    )
+
+    # If we are supplied a JSON realisation, we can add lobels for jump points.
```
```suggestion
    # If we are supplied a JSON realisation, we can add labels for jump points.
```

joelridden reviewed (commented):
Just a few things, only mentioned the typer app() and title in 1 file but does apply to a few of the scripts. Love the documentation

AndrewRidden-Harper reviewed (dismissed):
Fantastic documentation!

lispandfound commented:
Because of #7 I will also add some plot tests and slip module tests. I think hypothesis testing the slip functions makes sense, and then regression testing on plots seems sensible. It's actually a really good way to check that the SRF parser is working the way it did before because these scripts calculate a lot of complicated things from the SRF.

lispandfound commented:
I've made a few changes since the last PR that are worth looking at. Especially @joelridden since most of the changes relate to requests from him:

1. Addressed PR comments.
2. Changed the scripts to use the updated realisation API (which slightly changes how realisations are loaded). I have also moved the workflow imports to under the if statement that checks if a realisation is given. I did this so that people can use these scripts while the workflow is still under development to avoid needing to include possibly broken code in the dependencies that prevents the scripts from running.
3. Added tests for all the new modules and CLI scripts. The CLI scripts are tested to check that they give the same output as in the wiki, like the following example:
```python
def test_plot_srf():
    with tempfile.NamedTemporaryFile(suffix=".png") as output_path:
        plot_srf.plot_srf(SRF_FFP, Path(output_path.name))
        assert md5sum(PLOT_IMAGE_DIRECTORY / "srf_plot_example.png") == md5sum(
            output_path.name
        )
```
Where as the slip module tests are a really good example of how to use hypothesis tests 
```python
@given(
    t0=st.floats(0.1, 100),
    dt=st.floats(1e-6, 100),
    total_slip=st.floats(1, 1000),
)
def test_cosine_slip(t0: float, dt: float, total_slip: float):
    """Check that the cosine slip function is a well-behaved slip function.

    For the cosine slip function we must have:

    1. Slip is always non-negative,
    2. Slip is zero at t0 and t1,
    3. Slip function integrates to total slip over the interval [t0, t1].
    """
    t1 = t0 + dt
    t = np.linspace(t0, t1, num=20)
    slip_function = slip.cosine_slip(t, t0, t1, total_slip)

    # Slip should be non-negative and zero at boundaries
    assert np.all(slip_function >= 0)
    # Need to be a bit more lenient on boundary conditions because of
    # np.cos approximation errors.
    assert slip_function[0] == pytest.approx(0, abs=1e-6)
    assert slip_function[-1] == pytest.approx(0, abs=1e-6)
    # Total slip integrated should be close to the total slip given.
    # Using scipy quad because numpy trapz is not good enough
    total_slip_from_function, _ = sp.integrate.quad(
        lambda t: slip.cosine_slip(t, t0, t1, total_slip), t0, t1
    )
    # Need to be a bit more lenient for the approximation here.
    assert np.allclose(total_slip_from_function, total_slip, atol=1e-6)
 ```
 That is no test cases, just a description of what the parameters look like and the properties that should hold, and let hypothesis generate the rest. I think a good practice for non-trivial docstrings on test cases should be a plain English description of what the test checks (rather than the usual numpydoc-style docstrings we write for regular code). Since we will never actually call these test functions documenting every argument seems silly.
 
 Until #7 is merged, you'll just have to take my word that the test coverage is there and the tests pass.

claudio525 reviewed (dismissed):
Looks great!

lispandfound commented:
Ok this pull is now finally in a good state again. The tests now run because the other PR got merged, and I fixed the coverage issues. There are a couple `# pragma: no cover` lines in a couple of scripts to tell the coverage tool to ignore the fact that the realisation related code is not tested. This isn't tested because workflow is not ready yet. Consider the behaviour of those branches beta and subject to change.

AndrewRidden-Harper commented on source_modelling/slip.py:
```diff
@@ -26,13 +26,13 @@ def rise_time_from_moment(moment: npt.ArrayLike) -> npt.NDArray[np.floating]:
     Parameters
     ----------
-    t : float
+    t : array-like
```
Just to check, should this (and others) be array_like or is array-like ok?

AndrewRidden-Harper reviewed (commented):
It looks good. I just want to check that you intended to use array-like instead of array_like?

lispandfound commented:
@joelridden All good for you to review now.
